pandas/core/generic.py

```diff
@@ -2774,40 +2774,28 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
         # set the default here, so functions examining the signaure
         # can detect if something was set (e.g. in groupby) (GH9221)
         if axis is None:
-            axis = 0
+            axis = self._stat_axis_name
         axis = self._get_axis_number(axis)
-        method = mis._clean_fill_method(method)
+        method = missing._clean_fill_method(method)
 
         from pandas import DataFrame
         if value is None:
             if method is None:
                 raise ValueError('must specify a fill method or value')
-            if self._is_mixed_type and axis == 1:
+            if self._is_mixed_type:
+                if (self.ndim > 2) and (axis == 0):
+                    raise NotImplementedError('cannot fill across axis 0 for mixed dtypes')
```

Comment:
I don't recall that. my point is that this should be much simpler logic. Too many if statements here (and in the internals).